src/clj/puppetlabs/http/client/async.clj

```diff
@@ -117,21 +114,28 @@
                               query-params)]
       (.build uri-builder))))
 
-(defn- content-type
-  [{:keys [headers]}]
+(defn content-type
+  [body {:keys [headers]}]
   (if-let [content-type-value (some #(when (= "content-type"
                                            (clojure.string/lower-case (key %)))
                                        (val %))
                                     headers)]
-    (let [content-type (ContentType/parse content-type-value)]
-      (if (.getCharset content-type)
-        content-type
-        (ContentType/create (.getMimeType content-type) Consts/UTF_8)))))
+    ;; In the case when the caller provides the body as a string, and does not
+    ;; specify a charset, we choose one for them.  There will always be _some_
+    ;; charset used to encode the string, and in this case we choose UTF-8
+    ;; (instead of letting the underlying Apache HTTP client library
+    ;; choose ISO-8859-1) because UTF-8 is a more reasonable default.
+    (let [content-type (ContentType/parse content-type-value)
+          charset (.getCharset content-type)
+          should-choose-charset? (and (string? body) (not charset))]
+      (if should-choose-charset?
+        (ContentType/create (.getMimeType content-type) Consts/UTF_8)
+        content-type))))
```

Comment:
The code duplication is much bigger than just this section.  Ideally, I'd like to see this get to the point where most of the client configuration and handling is done in the Java layer and the Clojure-level API just uses the backend Java APIs rather than duplicating a lot of that functionality.  I swear I created a JIRA ticket to talk about that in particular but can't seem to find it now.  I can resubmit that, @cprice404, if you don't have any recollection of this.